### DISCUSSION TITLE
Importing / exporting content

### DISCUSSION THREAD

**Discussion by jmikrut** [payloadcms/payload#4309] [Roadmap]
We are planning to build an import / export function that can help export to Excel, CSV, etc.

This will include upload import / export as well.

**Comment by matthijs166** [comment]
In Wordpress we used wp-all-import/export plugin a lot. Love the way that plugin is build.

**Comment by willviles** [comment]
This would be great. Could this functionality be built with RBAC at collection level for both imports and exports?

**Comment by matthijs166** [comment]
**Here is a breakdown of features I really like about the export and import plugin:**
Multiple export types
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/d7b83ba8-b6a0-4b58-9147-0ca196b70857">
...

Filter options to exclude results in the export
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/eea84ce9-9110-499e-bf3e-e334c43f4fd9">
...

Mapping/selecting fields
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/7fec52a8-9eda-4356-a67b-b71f0b873c76">
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/ae94a96f-d489-4a3c-9526-2be78c640f16">
<img width="500" alt="Screenshot 2023-11-30 at 20 21 43" src="https://github.com/payloadcms/payload/assets/31346561/f509d4f4-5c13-4834-b7bf-540cbc84684b">
...

Reusability of running exports/imports later again. Setting up a nice export/import can take some time.
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/d5ada197-3335-4153-b609-009df0b892fc">
...

Setup for cron and having a URL for a feed. (handy if other software import the feed in an interval)
<img width="500" alt="Screenshot 2023-11-30 at 20 14 00" src="https://github.com/payloadcms/payload/assets/31346561/0a39571e-ba81-4315-baef-9bcb983c99a5">
...

Import from URL
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/4d95b8c1-6429-4621-986d-3e28632f99dc">
...

Record matching & field selection what to overwrite
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/ffd26323-e21a-4976-81e7-8ddc477c5907">
...

Custom code for data normalization and transformations
<img width="500" alt="image" src="https://github.com/payloadcms/payload/assets/31346561/52b26e10-e966-4cbe-a560-c52543836f94">
...

Those are my top features, probly to much for the first version 😓 
This feature would be very awesome to sync Dev/staging with production without exporting user data.

**Comment by JanettaH** [comment]
What is the status of this functionality?

**Comment by NicoCaldo** [comment]
Is there a plan for "import from Wordpress". I think if we were able to migrate a complete Wordpress blog this project would skyrocket.

**Comment by Metax7** [comment]
Does anyone know how to export data from WordPress to Payload?

**Reply by katiasmet-93** [reply]
I imported data from prismic. I exported it to a json file. Then created a small app to do a post fetch to the payload API. You could do something similar with Wordpress.

**Comment by danielkoller** [comment]
Any info concerning the export function? 

I would like to give my team the possibility to export form entries as a csv. 

Thanks!

**Comment by kalon-robson** [comment]
![image](https://github.com/user-attachments/assets/9c8116a3-1b56-4953-8947-115609034af1)

I've reopened this project that I started a while back

- This version only migrates between payload instances¹ ², which is really handy when we want to diagnose an issue on an existing website so we migrate the prod content to our localhost.
- The next version, I want to add mapping CSV columns to payload fields.
- Then it would be handy to selectively export information by reusing the payload filters (on the collections pages) and selectively picking what fields we want.

I'll prepare the plugin to public and share it here if anyone wants to help to develop this with me.

¹ Ideally the collections are identical, from my testing payload is pretty smart with ignoring extra bits of data when creating records
² There is no collision detection and is intended to be used to import into blank instances

**Reply by tempo22** [reply]
> ![image](https://private-user-images.githubusercontent.com/29440540/359985253-9c8116a3-1b56-4953-8947-115609034af1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3MDA3NTcsIm5iZiI6MTcyNzcwMDQ1NywicGF0aCI6Ii8yOTQ0MDU0MC8zNTk5ODUyNTMtOWM4MTE2YTMtMWI1Ni00OTUzLTg5NDctMTE1NjA5MDM0YWYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDEyNDczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyYTdiZDdjOTZlYzZmNDgzMWZkYjYwMTM0ZjkzODUxN2I2OWFlZWQ2NzMyMDE2NzI4ZjM3YWViMTczNzYwOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m3zhreq3lYFLX7X3_DXM1g2pdMg4Pu-nGL6H4v6kvgA)
> 
> I've reopened this project that I started a while back
> 
> * This version only migrates between payload instances¹ ², which is really handy when we want to diagnose an issue on an existing website so we migrate the prod content to our localhost.
> * The next version, I want to add mapping CSV columns to payload fields.
> * Then it would be handy to selectively export information by reusing the payload filters (on the collections pages) and selectively picking what fields we want.
> 
> I'll prepare the plugin to public and share it here if anyone wants to help to develop this with me.
> 
> ¹ Ideally the collections are identical, from my testing payload is pretty smart with ignoring extra bits of data when creating records ² There is no collision detection and is intended to be used to import into blank instances

Did you had the chance to prepare the public plugin ?

**Reply by kalon-robson** [reply]
Hi @tempo22, I'm in a bit of a pickle with the new upcoming 3.0 release.

This version uses express routes heavily to achieve it's goals and I'm not sure if I should finish it like it is and then do a migration to 3.0 using the next routes.

I'm guessing it would still be handy to maintain this in 2.0 as the adoption of 3.0 will take time. What do you think?

Anyhow thank you for bumping this to my attention. I will notify you with the repo soon.

**Comment by akhrarovsaid** [comment]
As it stands, the only way to get _true_ import/export is to build it yourself in-house with requirements from your specific project. I've yet to use a plugin that blew me away. It's definitely a much needed feature with several use-cases. I've been considering turning my own solution into a Payload plugin which features not only the ability to export/import from any specified collection, but also to choose what fields to export and how to reduce complicated fields such as Media/Upload and nested fields.

![Export-example10-01-2024_06_37_PM](https://github.com/user-attachments/assets/9b434eff-40b1-489f-94c3-701ebfb722f5)

This is being used _live_ right now for one of my clients and it works exceptionally well. However, it works well specifically because it's delegated to a **job** via a **queue.** This allows my customers the freedom of exporting/importing **tens of thousands** of entries without forcing the user to sit there and wait for it all to complete. Ideally, this wouldn't even happen on the same instance, but a dedicated server just for imports and exports.

I think before Imports/Exports really happen, what we **really** need is for the [Jobs/Queues feature](https://github.com/payloadcms/payload/pull/8228) to ship. This would be an absolute game changer. In the meantime, any solution will just leave much to be desired.

**Reply by DavideBecker** [reply]
This is something I'd be very interested to take a look at your approach to it, as we currently write our own import/export solution in-house as well

**Reply by danielkoller** [reply]
Could you please share how you do it? :)

**Reply by mumincelal** [reply]
@akhrarovsaid could you share your approach with us? We need a solution like this.

**Reply by matthijs166** [reply]
Oehh very cool!!

**Reply by akhrarovsaid** [reply]
@mumincelal I'll do you one better: since this seems to have attracted a lot of interest, I'll author a plugin for this the moment I find the time. First iteration without job queues to account for those users who enjoy deploying to serverless, and then I'll add an opt-in jobs variant behind a config option is what I'm thinking now that job queues have shipped.

**Reply by crazytweeks** [reply]
Looking forward for this implementation for payload 3.

**Reply by BrianJM** [reply]
@akhrarovsaid Did you release this?

**Reply by marwinhormiz** [reply]
Any updates on this?

**Reply by dankhazipeter** [reply]
any update?

**Comment by AdvithShetty** [comment]
Hi, Any Update on this feature? is there any ETA or alternative to export collections untill its released?

**Reply by jhb-dev** [reply]
It looks like an import/export plugin is currently in development: https://github.com/payloadcms/payload/pull/10795

**Reply by AdvithShetty** [reply]
Yes i did see that one but, Any idea on the ETA?

**Reply by DracoBlue** [reply]
was published with 3.27.0 https://github.com/payloadcms/payload/releases/tag/v3.27.0

**Comment by lorenzrabs** [comment]
Hey, great work so far!

My two main USPs for this feature are:

- Exporting all website content to send it to a translation agency and then importing the completed translation back.
- Exporting all content (essentially like a database dump) to import it onto another domain—so that only the media files need to be transferred.

Would these functions be possible in future releases?

**Comment by matthijs166** [comment]
Just tried it and this feature is awesome! Can't wait on the import functionality of this package

**Comment by JW148** [comment]
Hi, is this plugin available as an NPM package yet? As mentioned above it is included in the 3.27.0 release, but I can't seem to find it anywhere.

**Reply by DracoBlue** [reply]
It is called

   @payloadcms/plugin-import-export

**Reply by JW148** [reply]
Awesome, thanks! Must have been a typo on my end

**Reply by notflip** [reply]
It seems like there's no documentation for this yet? Or rather it's a 404: https://payloadcms.com/docs/plugins/import-export

**Comment by matthijs166** [comment]
We played a bit around with the plugin and it is awesome!

The features we are missing are mapping relations to a single values.
For example, exporting the collection with registrations and outputting the name (instead of the object ID).
Would also be awesome to map multiple fields, such as a name and an email.

Docs about the plugin would also help a lot.

Just an idea:
Sometimes we also need XML feeds for Google shopping and the sitemap.
An option for XML formats would also be nice, and maybe even with the option to predefine the structure with export templates.
(shipping the sitemap by default with the plugin)

We try to contribute to the plugin, If we got some extra time on our hands!

**Reply by notflip** [reply]
Would you mind pasting a gist of the code you're using? Not having documentation is a big issue at the moment.

**Comment by 2coo** [comment]
I tried it but it shows error and It doesn't show any data in preview

```
err: {
      "type": "APIError",
      "message": "Collection with slug  not found",
      "stack":
          APIError: Collection with slug  not found
          at createExport
```


my plugins config.

```
importExportPlugin({
    collections: [
      "news"
    ],
    overrideExportCollection: (collection) => {
      console.log("collection", collection);
      collection.admin.group = "System"
      return collection
    },
    disableJobsQueue: true,
  }),
  ```

**Reply by DracoBlue** [reply]
Try without the overrideCollection part.

**Reply by renjurajt** [reply]
I am using 3.33.0. The functionality doesnt work.

**Comment by Mondonno** [comment]
Is it working? The pull was on March 4: https://github.com/payloadcms/payload/pull/10795

**Reply by danielkoller** [reply]
Yes, it does. Just installed and played around.

**Reply by orhancelik1** [reply]
> Yes, it does. Just installed and played around.

How you download exports when deploy in vercel? it is working only in my local machine. Do you have any blob or s3 setup for this?

**Comment by AlexRoosWork** [comment]
Great job on this plugin! 
I am also missing some documentation for how to set it up as pointed out by @notflip the documentation page currently returns a 404. 

Nevertheless, by following the comments in this thread I was able to get the export to work! 💪

Now, I understand that the import functionality of this plugin is not yet implemented. 

**Is there any ETA on when the import feature will become available?**

**Reply by DanRibbens** [reply]
I am working on the plugin docs: https://github.com/payloadcms/payload/pull/13243

There is a little more detail that I want to add about how the export can be used via admin UI or how to call it in code or with the jobs queue.

Let me know if you have any specific questions that should go in.

Thanks!

**Reply by SebastianSchuetze** [reply]
@DanRibbens Any plans to make the disable options for the button and save more dynamic via checking user roles or certain attributes like the access in collections?

I already have the requirements to only allow admin role based users to do that but not normal user accessing admin ui.

**Reply by DanRibbens** [reply]
> @DanRibbens Any plans to make the disable options for the button and save more dynamic via checking user roles or certain attributes like the access in collections?
> 
> I already have the requirements to only allow admin role based users to do that but not normal user accessing admin ui.

I would consider it a bug if the Save button is not respecting the access control for the `exports` collection if you have extended that in your project. 

The download button is a different case. I consider it a feature that download is available for users regardless of whether they can create a file. Because of this separation of functionality it seems to me that we need to extend the `disableDownload` type to include a function with the `req` so you can return booleans based on conditions off your `user`.

What do you think of this approach?

Contributions here would be very welcome as I don't have a timeline as to when we can get to this.

**Comment by danielkoller** [comment]
Hi @DanRibbens, great work so far and thanks for updating the README.

I'm implementing the plugin for easier translations and need help with a specific use case. I'd like to export a page's structure showing only the blocks that are actually used, along with their localized content. This would help my editors easily identify what still needs translation and where it's located.

**Current issue:** When I add the layout field to the export, I get all available blocks rather than just the ones actually used on the page.

**Requirements:**

- Only blocks that are actually used on the page
- Block name + content (simplified - I don't need the full RichText structure)
- Localized content to show translation status

**Question:** Is this functionality already supported by the current plugin, or would I need to implement a custom solution?
Thanks again for the excellent work!

**Comment by giorgiPapava** [comment]
Is it possible to export/import globals with this plugin?

**Comment by rafaell-lycan** [comment]
Any news on this feature? 👀

**Comment by magicspon** [comment]
CraftCMS has a really handy feature where you can export the database from within the control panel. This would make a great addition to payloadcms. At the moment I'm not clear what the best approach is for pulling down a production database into a local environment.

**Reply by DracoBlue** [reply]
For this I would suggest pulling the whole database and importing it via sql. That is a consistent backup and takes care of all foreign keys and the likes.

**Comment by sebastijandumancic** [comment]
Hey @jmikrut Is there any roadmap plan on import feature? We will need this very soon so I'd build it myselft if it's not coming in next month. Is there any ETA? Thanks for the great work!

**Reply by saroroce** [reply]
man chill

**Reply by saroroce** [reply]
https://github.com/saroroce/payload-extended-import-export-plugin

try this plugin

**Comment by saroroce** [comment]
guys i have better version of import plugin but i need a time to make it like plugin to share with you

but for now for easy operations it works good https://github.com/saroroce/payload-extended-import-export-plugin

**Reply by SebastianSchuetze** [reply]
the plugin only talks about import. Does this include also export?

**Reply by saroroce** [reply]
export you need to use from payload export import plugin

**Comment by GermanJablo** [comment]
[Our plugin](https://payloadcms.com/docs/plugins/import-export) was released quite some time ago and several of the suggestions mentioned here have already been implemented. To keep this more traceable, please open new, more specific feature requests. Thank you!
